b2sdk/scan/folder.py

```diff
@@ -178,15 +178,43 @@ def ensure_non_empty(self):
             raise EmptyDirectory(self.root)
 
     def _walk_relative_paths(
-        self, local_dir: str, relative_dir_path: str, reporter,
-        policies_manager: ScanPoliciesManager
+        self,
+        local_dir: str,
+        relative_dir_path: str,
+        reporter,
+        policies_manager: ScanPoliciesManager,
+        visited_symlinks=None,
+        starting_dir: str = None,
     ):
         """
         Yield a File object for each of the files anywhere under this folder, in the
         order they would appear in B2, unless the path is excluded by policies manager.
 
         :param relative_dir_path: the path of this dir relative to the scan point, or '' if at scan point
         """
+
+        def is_ancestor_directory(potential_ancestor, dir):
+            """
+            Helper function to check if the ancestor directory is an ancestor
+            of the descendant directory.
+
+            :param potential_ancestor: The ancestor directory path to check.
+            :param dir: The directory to check ancestor against.
+            :return: True if the ancestor directory is an ancestor of the descendant directory, False otherwise.
+            """
+
+            # should return true if ancestor is located anywhere in the path of descendant
+            # e.g. /home/user/ancestor is an ancestor of /home/user/ancestor/descendant
+
+            potential_ancestor = os.path.abspath(os.path.realpath(potential_ancestor))
+            dir = os.path.abspath(os.path.realpath(dir))
+
+            return os.path.commonpath([potential_ancestor]) == os.path.commonpath(
+                [potential_ancestor, dir]
+            )
+
+        #FIXME should I construct starting_dir dynamically based on relative_dir_path and local_dir?
```

Comment:
@ppolewicz This parameter represents the original starting point of the directory scan. It is required to catch a situation described by @zackse in his comment here: https://github.com/Backblaze/B2_Command_Line_Tool/issues/513#issuecomment-426751216

While scanning the directory tree from the initial scanning point, we may jump outside of the original initial scanning point via symlinks. If the linked/target directory contains the initial scanning point, then we consider it a bad link.

Maybe this is not the correct implementation... After your comment, I am now thinking that maybe we should only rely on symlink visit counter to skip symlinks. In the current implementation I mixed the two approaches:

* I skip symlinks which are pointing to ancestor of the initial scan point
* I skip symlinks if we reach the symlink visist limit

I guess the second one is more general and covers the first one as a special case of sorts. But it will take more time to discover that we are actually in a circular symlinks case. Do you have any suggestions on what would be a better approach?

Reply:
refactored and removed starting_dir parameter completely